MM-29987 Implement new collapsed threads API

reflog opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
APIs implemented according to https://hackmd.io/xq00VeBsQ-SYDRPadW8yGw?view

API documentation PR: https://github.com/mattermost/mattermost-api-reference/pull/585
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-29987

#### Release Note
<!--
If no release notes are required, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your past-tense release note in the block below.
-->
```release-note
Added new API endpoints for working with experimental collapsed threads
```

esethna commented:
@reflog should we also add @enahum and @sudheerDev to this PR for review and transparency?

enahum commented:
LGTM, did not review in full as my golang skills died :P as long as everything matches the implementation doc we should be good

jespino commented on api4/user.go:
```diff
@@ -2806,3 +2815,225 @@ func migrateAuthToSaml(c *Context, w http.ResponseWriter, r *http.Request) {
+	w.Write([]byte(threads.ToJson()))
+}
+
+func markAsUnreadThreadByUser(c *Context, w http.ResponseWriter, r *http.Request) {
```
The mark as unread, at least needs a post or a date to set for the LastViewedAt, otherwise you only have the ability to mark as read or unread the whole thread.

jespino commented on store/sqlstore/post_store.go:
```diff
@@ -1937,11 +1934,13 @@ func (s *SqlPostStore) cleanupThreads(postId, rootId, userId string) error {
-	}
-	if _, err := s.GetMaster().Exec("DELETE FROM ThreadMemberships WHERE PostId = :Id", map[string]interface{}{"Id": postId}); err != nil {
-		return errors.Wrap(err, "failed to delete ThreadMemberships")
+	if permanent {
```
Maybe move this up to avoid update and remove.

jespino commented on store/sqlstore/thread_store.go:
```diff
@@ -108,6 +108,144 @@ func (s *SqlThreadStore) Get(id string) (*model.Thread, error) {
+}
+
+func (s *SqlThreadStore) Follow(userId, threadId string, state bool) error {
+	query, args, _ := s.getQueryBuilder().
```
What happens if the ThreadMembership doesn't exists yet?

reflog commented:
@jespino all great points. updated the API to support timestamp for read state, addressed your other points. related API documentation PR is updated too.

jespino commented on store/sqlstore/thread_store.go:
```diff
@@ -203,29 +203,16 @@ func (s *SqlThreadStore) GetThreadsForUser(userId string, opts model.GetUserThre
-	}
-	query, args, _ := qb.ToSql()
+func (s *SqlThreadStore) MarkAllAsRead(userId string, timestamp int64) error {
+	query, args, _ := s.getQueryBuilder().Update("ThreadMemberships").Where(sq.Eq{"UserId": userId}).Set("LastViewed", model.GetMillis()).ToSql()
```
This is not using the timestmap

reflog replied:
facepalm

jespino commented on web/params.go:
```diff
@@ -236,6 +237,12 @@ func ParamsFromRequest(r *http.Request) *Params {
 	}
 
+	if val, err := strconv.ParseInt(query.Get("timestamp"), 10, 64); err != nil || val < 0 {
+		params.Timestamp = model.GetMillis()
```
This can lead on a not easy to detect bug around timesmap, for example, passing a wrongly formatted timestamp gets automatically converted in to "now"

reflog replied:
Mkay. I'll add strict checking

jespino reviewed (approved):
LGTM

ashishbhate commented on api4/user_test.go:
```diff
@@ -5255,3 +5255,308 @@ func TestUpdatePassword(t *testing.T) {
+
+		var rootIds []*model.Post
+		for i := 0; i < 30; i++ {
+			time.Sleep(1)
```
Why is the sleep here (and on line 5382) required?

reflog replied:
@ashishbhate I think I added it to get different timestamps on posts

ashishbhate reviewed (approved):
LGTM! 

Just one non-blocking question.

esethna commented:
@reflog can we clarify what is needed from QA on this PR?

reflog commented:
@esethna - from QA perspective - nothing to check yet, since this is just the API implementation and it's not used anywhere at the moment. Once the UI is hooked up to it - then the QA testing can being. Unless we are doing API level QA testing, which I don't think we do
